proposed/0006-Launcher.md

```diff
@@ -0,0 +1,204 @@
+
+**/path/to/app/binary --launcher "udp:<random-utf8-port>:<random-utf8-string>"**
+
+All parameters are utf8 Strings.
+
+**step 4:** Launcher will wait for a predefined time of **15** seconds for data reception on that port. If it times out it will close the socket (release its binding to it)
+
+**step 5:** App generates a random encryption asymmetric keypair - `<App-Asymm-Keys>`. Then it responds on the socket asking for Launcher to give it an `<App-Specific-Symm-Key>`, its root directory-key and SAFEDrive directory-key, which Launcher had reserved as `XYZ-Root-Dir`
+- The payload format for this request shall be CBOR encoded sturcture of the following:
+```
+struct Request {
+    launcher_string      : String, // This must be the one supplied by Launcher
+    nonce                : sodiumoxide::crypto::box_::Nonce,
+    public_encryption_key: sodiumoxide::crypto::box_::PublicKey, // from <App-Asymm-Keys>
+}
+```
+
+**step 6:** The Launcher verifies the `launcher_string` field above and generates a strong random symmetric encryption key `<App-Specific-Symm-Key>`. This is encrypted using App's `public_encrytion_key` and `nonce` above.
+
+**step 7:** Launcher gives the App what it requested.
+- The payload format for this response shall be CBOR encoded sturcture of the following,
+```
+struct Response {
+    cipher_text       : Vec<u8>, // encrypted symmetric keys
+    app_root_dir_key  : DirectoryKey,
+    safe_drive_dir_key: Option<DirectoryKey>,
+    tcp_listening_port: u16, // <Laucher-Port>
+}
+
+Where DirectoryKey consists of NameType, u64 (tag), AccessLevel and bool (versioned or not) as defined in safe_nfs crate.
+```
+- At this point Launcher closes the UDP socket (ie., is no longer bound to it).
+
+## Reads and Mutations by the App
+
+**step 0:** App will connect to the Launcher Background process using `<Launcher-IP>:<Launcher-Port>`.
+
+**step 1:** All `GET/PUT/POST/DELETE`s will go via the Launcher. The App will essentially encrypt using `<App-Specific-Symm-Key>` and pass it to Launcher Background Process which will decrypt and Re-encrypt and sign it using normal process (currently MAID-keys). For `GET`s it will be the reverse - Launcher will eventually encrypt using `<App-Specific-Symm-Key>` before handing the data over to the App.
+
+**step 2:** Since `<App-Specific-Symm-Key>` is recognised by Launcher only for current session, there is no security risk and the App will not be able to trick Launcher the next time it starts to use the previous keys to mutate network or read data on its behalf.
+
+## Share Directory App Flow
+
+Every time the App tries to access `SAFEDrive` Launcher will check the permission in `<MAIDSAFE-SPECIFIC-CONFIG-ROOT>/LauncherReservedDirectory/LauncherConfigurationFile`.
+### Grant Access
```

Comment:
We could probably combine grant and revoke points into a single one.  Also we should probably specify the level of access the app gets (i.e. R/W).  Something like:

"User can grant the app read/write access to the `SAFEDrive` directory or revoke that access by adding or removing ... respectively"